lib/gringotts/gateways/pin_payments.ex

```diff
@@ -0,0 +1,244 @@
+
+    card_token_response =
+      :post
+      |> commit("cards", card_for_token(card, opts) ++ Keyword.delete(opts, :address))
+      |> extract_card_token
+
+    case card_token_response do
+      {:error, error} ->
+        {:error, error}
+
+      {:ok, token} ->
+        params =
+          [
+            amount: value,
+            capture: false,
+            card_token: token,
+            currency: currency
+          ] ++ Keyword.delete(opts, :address)
+
+        commit(:post, "charges", params)
+    end
+  end
+
+  def authorize(amount, card_token, opts) when is_binary(card_token) do
+    {currency, value, _} = Money.to_integer(amount)
+
+    params =
+      [
+        amount: value,
+        capture: false,
+        currency: currency,
+        card_token: card_token
+      ] ++ Keyword.delete(opts, :address)
+
+    commit(:post, "charges", params)
+  end
+
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+
+  # Makes the request to PinPay's network.
+  # For consistency with other gateway implementations, make your (final)
+  # network request in here, and parse it using another private method called
+  # `respond`.
+
+  defp card_for_token(card, opts) do
+    [
+      number: card.number,
+      name: CreditCard.full_name(card),
+      expiry_month: card.month |> Integer.to_string() |> String.pad_leading(2, "0"),
+      expiry_year: card.year |> Integer.to_string(),
+      cvc: card.verification_code,
+      address_line1: opts[:address].street1,
+      address_city: opts[:address].city,
+      address_country: opts[:address].country,
+      address_line2: opts[:address].street2,
+      address_postcode: opts[:address].postal_code,
+      address_state: opts[:address].region
+    ]
+  end
+
+  @spec commit(atom, String.t(), keyword) :: {:ok | :error, Response}
+  defp commit(:post, endpoint, param) do
+    auth_token = encoded_credentials(param[:config].apiKey)
+
+    headers = [
+      {"Content-Type", "application/x-www-form-urlencoded"},
+      {"Authorization", auth_token}
+    ]
+
+    url = @test_url <> "#{endpoint}"
+    param = Keyword.delete(param, :config)
+
+    url
+    |> HTTPoison.post({:form, param}, headers)
+    |> respond
+  end
+
+  defp encoded_credentials(login) do
+    hash = Base.encode64("#{login}:")
+    "Basic #{hash}"
+  end
+
+  defp extract_card_token({:ok, %{token: token}}) do
+    {:ok, token}
+  end
+
+  defp extract_card_token({:error, error_response}) do
+    {:error, error_response}
+  end
+
+  # Parses PinPay's response and returns a `Gringotts.Response` struct
+  # in a `:ok`, `:error` tuple.
+  @spec respond(term) :: {:ok | :error, Response}
+
+  defp respond({:ok, %{status_code: code, body: body}}) when code in 200..299 do
+    {:ok, parsed} = decode(body)
+    token = parsed["response"]["token"]
```

Comment:
This is wrong!! Please read the `Response.t` docs, the `token` field is NOT for the _authorization token ID_ -- the auth token ID should go into `Response.id` field!
```elixir
token = parsed["response"]["card"]["token"]
```